src/client_side_request.cc

```diff
@@ -2022,6 +2022,25 @@ ClientHttpRequest::handleAdaptedHeader(Http::Message *msg)
     } else if (HttpReply *new_rep = dynamic_cast<HttpReply*>(msg)) {
         debugs(85,3,HERE << "REQMOD reply is HTTP reply");
 
+        if (request->method == Http::METHOD_CONNECT) {
+            // CONNECT requests that REQMOD produces a reply for need special
+            // handling.
+            if (Http::Is2xx(new_rep->sline.status())) {
+                // REQMOD replying with a 2xx is a bug, since Squid won't have
+                // connected to the origin server.
+                debugs(85, DBG_IMPORTANT, HERE << "REQMOD produced a 2xx HTTP response for a CONNECT request");
```

Comment:
Apologies for leaving my PRs alone for a while - I've been stuck with some other work and expect to return to them soon.  However, here are some answers:

### Treating 2xx responses in the same way as any other response:

If we treat a 2xx response in the same way as any other response here, the 2xx response will be forwarded to the client, and then (in the case of keep-alive connections) Squid will wait for a new HTTP request.

This is a broken situation - the client has been told that it now has a CONNECT tunnel to an origin server, whilst Squid is expecting another HTTP request which is unrelated to the CONNECT.  The client is probably trying to access an HTTPS URI, so it will send a TLS Client Hello instead of the HTTP request that Squid was expecting.

I decided that it would be better to produce an error rather than losing synchronisation between Squid and the client.

### What could we do with 2xx responses?

Firstly, without adaptation meddling with the CONNECT, the connection would potentially be handled by ssl bump anyway.  So in order to bump the connection and then have adaptation handle the HTTP requests within it, we wouldn't need to do any adaptation on the CONNECT request itself.

Therefore, my presumption is that the only reason why adaptation might to produce a 2xx HTTP response for a CONNECT is to allow access to the raw data stream, rather than to facilitate any SSL bump related things. (*)

At the point that REQMOD adaptation has generated a 2xx response, Squid has not established a tunnel to anywhere, but we've told the client that one has been established.  So what _could_ we do?

We could establish a tunnel to somewhere.  But there are a couple of problems:
1. Where would we establish the tunnel to? There is no information in the ICAP response indicating where we should connect.
2. If we try to establish a tunnel but it fails, we're in trouble because we already told the client it had succeeded.
Note that instead of generating an HTTP reply as we are doing here, REQMOD can replace the HTTP request with a new CONNECT and allow Squid's normal logic to attempt to establish the tunnel.

Alternatively, we could connect the tunnel to the adaptation service.  Squid doesn't currently have any code to allow an adaptation service to see, process or generate the raw data within the CONNECT tunnel itself.  In the future it would be nice for adaptation to be able to handle raw tunnelled data (both for CONNECT and things like WebSockets), but there is no support for that at the moment, and I expect the ICAP protocol would need to be extended to cope with bidirectional data streams.

So in short, whilst we _could_ make Squid do something useful when ICAP generates a 2xx HTTP response to a CONNECT, it is not at all clear what that useful thing might be.  Whatever useful thing is decided upon, it would certainly require a lot of work to implement.

(*) Alternatively, we could trigger client-first bumping and then insist that all requests within the bumped tunnel result in REQMOD generating an HTTP reply since there we're not connected to any real origin server.  But as mentioned, you can already achieve this in a different way so I don't see a lot of value in it.

Reply:
I agree with most of your points, but they do not imply that an ICAP service satisfying CONNECT with a `200 Connection Established` response is _buggy_. It may not be compatible with the current Squid code, but that is not a Squid bug or a service bug.

Furthermore, I wonder whether there are special use cases that will work here if we allow them. I am thinking about a situation where, for example:

* the client is establishing a CONNECT tunnel to a plain HTTP-speaking[^1] (no TLS) server on port 443, and
* an ICAP service that knows about that client and wants to intercept/satisfy its messages.
* Squid allows but does not SslBump this CONNECT tunnel.

[^1]: Or it could just look like plain HTTP to Squid but in fact it is a different protocol on top of HTTP that the ICAP service is aware of.

AFAICT, at least on the surface of things, Squid can handle that situation without extra code/features -- all the post-CONNECT traffic will be HTTP requests that Squid will parse and forward to the ICAP REQMOD service, and the ICAP service will _satisfy_ them.

If the above hypothetical use case works, the question becomes: Why ban that (and possibly other) use cases? The answer could be that most adaptation services that satisfy CONNECT requests are probably broken/incompatible and, if we allow this use case, we will significantly complicate triage of those cases. Thus, we block all 2xx responses and wait for somebody with genuine need to add a configuration option to allow them. Is that our rationale here?

Reply:
> I agree with most of your points, but they do not imply that an ICAP service satisfying CONNECT with a 200 Connection Established response is buggy. It may not be compatible with the current Squid code, but that is not a Squid bug or a service bug.

There is no specification that I'm aware of that specifies what should actually happen where an ICAP service produces a "200 Connection Established" response (in terms of handling the connection after the response it returned to the client).  So any ICAP service that does this is now outside the realms of specified behaviour - you have to know how the specific ICAP client will handle that response.  Maybe "buggy" is too strong a term, but it is undefined behaviour (which may be defined in the future, and the eventual definition may not match what we implement :).

> AFAICT, at least on the surface of things, Squid can handle that situation without extra code/features -- all the post-CONNECT traffic will be HTTP requests that Squid will parse and forward to the ICAP REQMOD service, and the ICAP service will satisfy them.

I don't believe that would work, because the client would be making origin server HTTP requests instead of proxy HTTP requests.  i.e. the requests would look like "GET /" instead of "GET http://example.com/".

We could certainly make Squid handle these requests (in particular, it would have to make an assumption about the URI scheme), but I don't believe it will work out of the box.

I agree that producing an error helps triage ICAP services which are producing a response which results in undefined behaviour.  I don't believe that allowing 2xx replies is of much use without further modifications to squid, but I am happy to remove the ban on 2xx replies if that's the preferred way forward.

Reply:
Thank you for a thoughtful response. I have doubts about some of the assertions in your statements, but I am OK with banning these 2xx responses (at least for now) as long as we do not label them as bugs.

I suggest giving the level-1 message an ID, so that admins can control the message frequency via cache_log_message -- the affected ICAP service is likely to emit many such messages.

Do you want to take a stub at the corresponding changes?

* If yes, please also drop the deprecated `HERE`. I would also use a "REQMOD satisfied CONNECT with an HTTP 2xx response" wording to emphasize that we are dealing with request satisfaction here. Technically, a REQMOD service can modify origin server responses as well (but Squid does not support post-cache REQMOD yet).
* If no, I can help with these adjustments.

Reply:
I'm happy to implement the changes.  It won't be this week though I'm afraid.

Reply:
Thank you. No rush at all!